services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloaderSubscriber.java

```diff
@@ -27,86 +27,115 @@
+     * copy has the partNumber field updated as more parts are downloaded.
+     */
     private final GetObjectRequest getObjectRequest;
 
-    private AtomicBoolean totalPartKnown = new AtomicBoolean(false);
-    private AtomicInteger totalParts = new AtomicInteger();
-    private AtomicInteger completed = new AtomicInteger(0);
-    private AtomicInteger currentPart = new AtomicInteger(0);
+    /**
+     * Set to true when we know the total part number of the object to get.
+     */
+    private final AtomicBoolean totalPartKnown = new AtomicBoolean(false);
+
+    /**
+     * Once total part is know, this value indicates the total number of parts of the object to get.
+     */
+    private final AtomicInteger totalParts = new AtomicInteger();
+
+    /**
+     * The total number of completed parts. A part is considered complete once the completable future associated with its request
+     * completes successfully.
+     */
+    private final AtomicInteger completedParts = new AtomicInteger(0);
 
+    /**
+     * The subscription received from the publisher this subscriber subscribes to.
+     */
     private Subscription subscription;
 
-    private CompletableFuture<GetObjectResponse> responseFuture;
-    private GetObjectResponse returnResponse;
+    /**
+     * This future will be completed once this subscriber reaches a terminal state, failed or successfully, and will be
+     * completed accordingly.
+     */
+    private CompletableFuture<Void> future = new CompletableFuture<>();
 
     public MultipartDownloaderSubscriber(S3AsyncClient s3, GetObjectRequest getObjectRequest) {
         this.s3 = s3;
         this.getObjectRequest = getObjectRequest;
-        this.responseFuture = new CompletableFuture<>();
     }
 
     @Override
     public void onSubscribe(Subscription s) {
-        if (this.subscription == null) {
-            this.subscription = s;
+        if (this.subscription != null) {
+            s.cancel();
+            return;
         }
-        s.request(1);
+        this.subscription = s;
+        this.subscription.request(1);
     }
 
     @Override
     public void onNext(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
-        int part = currentPart.incrementAndGet();
-        if (totalPartKnown.get()) {
-            log.info(() -> String.format("total part: %s, current part: %s", totalParts.get(), part));
-        } else {
-            log.info(() -> String.format("part %s", part));
+        if (asyncResponseTransformer == null) {
+            throw new NullPointerException("onNext must not be called with null asyncResponseTransformer");
         }
-        if (totalPartKnown.get() && part > totalParts.get()) {
-            log.info(() -> "no more parts available, stopping");
+
+        int nextPartToGet = completedParts.get() + 1;
+        if (totalPartKnown.get() && nextPartToGet > totalParts.get()) {
```

Comment:
As we discussed offline, let's move it down; cancel it as soon as we receive the last part.

Reply:
This was not working, as we might be cancelling while data is still being read from the response, which would lead to missing data in the body